proposals/0046-astro-env.md

```diff
@@ -0,0 +1,573 @@
+
+Those validators are also used at runtime by `getSecret`, more on that in a section below.
+
+## Client variables
+
+Client variables are made available through the `astro:env/client` virtual module.
+
+### Public client variables
+
+Public client variables, once validated, are directly exported from `astro:env/client`. Here is an example of the types generated and usage:
+
+```ts
+// .astro/env.d.ts
+declare module "astro:env/client" {
+  export const PUBLIC_STRIPE_KEY: string;
+}
+
+// Any .ts loaded by Vite
+import { PUBLIC_STRIPE_KEY } from "astro:env/client";
+```
+
+### Secret client variables
+
+This kind of variable is not supported as there's no safe way to send this data to the client. Types are made in such a way that TypeScript will not allow it nor the Astro config schema.
+
+## Server variables
+
+Server variables are made available through the `astro:env/server` virtual module. Importing this module client-side will throw an `AstroError`.
+
+### Public server variables
+
+Public server variables, once validated, are directly exported from `astro:env/server`. Here is an example of the types generated and usage:
+
+```ts
+// .astro/env.d.ts
+declare module "astro:env/server" {
+  export const API_PORT: number;
+}
+
+// Any .ts loaded by Vite
+import { API_PORT } from "astro:env/server";
+```
+
+### Secret server variables
+
+#### User API
+
+Secret server variables are available through `getSecret` exported from `astro:env/server`. This function allows retrieving a variable by its key:
+
+- If the key has not been specified in the schema, the function returns `string | undefined`
+- If the key is part of the schema, it will be validated at runtime using the custom validators and return the right data type. Note that an invalid variable will throw an `AstroError` at runtime only
+
+```ts
+// .astro/env.d.ts
+declare module "astro:env/server" {
+  type SecretValues = {
+    API_PORT: number;
+  };
+
+  type SecretValue = keyof SecretValues;
+
+  type Loose<T> = T | (string & {});
+  type Strictify<T extends string> = T extends `${infer _}` ? T : never;
+
+  export const getSecret: <TKey extends Loose<SecretValue>>(
+    key: TKey
+  ) => TKey extends Strictify<SecretValue>
+    ? SecretValues[TKey]
+    : string | undefined;
+}
+
+// Any .ts loaded by Vite
+import { getSecret } from "astro:env/dynamic";
+
+const API_PORT = getSecret("API_PORT"); // number
+const UNKNOWN_VARIABLE = getSecret("UNKNOWN_VARIABLE"); // string | undefined
+```
+
+#### Behaviors
+
+`getSecret` works differently depending on the case:
+
+- In dev mode, it uses `process.env`
+- While building (eg. static site or prerendered pages), it also uses `process.env`
+- In SSR, it throws unless an adapter provides its own implementation
+
+#### Adapters API
+
+##### `app.setGetEnv`
+
+Under the hood, `getSecret` uses `getEnv` from the runtime. It can be overrided by calling `app.setGetEnv`:
```

Comment:
Why make this part of the `App` class? The `App` class is for rendering. Setting environment variables is unrelated to rendering for my hosts.

In addition, doing it this way means that variables can only be accessed inside of a render function. If you try to access a variable in the top-level of a module like so:

```js
import { getSecret } from 'astro:env/server';

const API_URL = getSecret('API_URL');
```

This will result in the variable being `undefined` because `setGetEnv` is not called until the very end of the bundle.

Reply:
Initially I added `getEnv` to `app.render` options to have it per request but then moved it to a method. I thought calling `app.setGetEnv` right after the instantiation of `App` would be enough, where should that go then?

Reply:
I suppose we don't need the function `app.setGetEnv`, and just expose the function `setGetEnv` from `astro/env`.

Users can still call `setGetEnv` from anywhere inside their code, even during the rendering phase, and at the top level module.

Reply:
> I suppose we don't need the function app.setGetEnv, and just expose the function setGetEnv from astro/env.

Alright so we can't use `astro/env` as it's already taken by a dts (https://github.com/withastro/astro/blob/main/packages/astro/package.json#L34), what export should we use then?

> Users can still call setGetEnv from anywhere inside their code, even during the rendering phase, and at the top level module.

As for matthew's example, it will still fail for Cloudflare I believe but I guess it's just a Cloudflare limitation in general.

> Users can still call setGetEnv from anywhere inside their code, even during the rendering phase, and at the top level module.

Currently, I'm calling `setGetEnv` with an implementation that throws in the `App` constructor as I assumed it would run before anything else. Where should I put it instead?

Reply:
Yes, top-level usage is probably not possible in Cloudflare. That's just a limitation they've chosen, and shouldn't affect other runtimes.

As for where this lives, since adapters are built through Vite, is there any reason they can't just use a `astro:env` export? `astro:env/setup` or something.

Reply:
Sounds good

Reply:
The only issue I see is typing it, I guess it should not be done through typegen?

Reply:
I'm pretty sure you can add types for this, we have them in `packages/astro/client.d.ts`